packages/astro/src/runtime/server/index.ts

```diff
@@ -308,20 +311,36 @@ If you're still stuck, please open an issue on GitHub or join us at https://astr
 
 	// Rather than appending this inline in the page, puts this into the `result.scripts` set that will be appended to the head.
 	// INVESTIGATE: This will likely be a problem in streaming because the `<head>` will be gone at this point.
-	result.scripts.add(
-		await generateHydrateScript(
-			{ renderer: renderer!, result, astroId, props },
-			metadata as Required<AstroComponentMetadata>
-		)
+	const island = await generateHydrateScript(
+		{ renderer: renderer!, result, astroId, props },
+		metadata as Required<AstroComponentMetadata>
 	);
+	result._metadata.needsHydrationStyles = true;
 
 	// Render a template if no fragment is provided.
 	const needsAstroTemplate = children && !/<\/?astro-fragment\>/.test(html);
 	const template = needsAstroTemplate ? `<template data-astro-template>${children}</template>` : '';
+
+	if(needsAstroTemplate) {
+		island.props.tmpl = '';
+	}
+
+	island.children = `${
+		html ?? ''
+	}${template}`;
+
+	// Add the astro-island definition only once. Since the SSRResult object
+	// is scoped to a page renderer we can use it as a key to know if the script
+	// has been rendered or not.
+	let script = '';
+	if(!hydrationedRenders.has(result)) {
```

Comment:
```suggestion
	if(!hydratedRenders.has(result)) {
```

nit: should this be `hydrated` or `hydration` maybe?

Reply:
That was a weird name huh, I've updated it to something better.